src/library/tactic/simp_lemmas.cpp

```diff
@@ -864,10 +870,15 @@ static simp_lemmas add_core(type_context_old & ctx, simp_lemmas const & s, name
             proof = mk_app(proof, mvar);
         }
         expr lhs, rhs;
-        lean_verify(is_eq(type, lhs, rhs));
+        lean_verify(is_eq(type, lhs, rhs) || is_iff(type, lhs, rhs));
         if (symm) {
-            proof = mk_eq_symm(ctx, proof);
-            std::swap(lhs, rhs);
+            if (is_eq(type, lhs, rhs)) {
+                proof = mk_eq_symm(ctx, proof);
+                std::swap(lhs, rhs);
+            }
+        }
+        if (is_iff(type, lhs, rhs)) {
+            proof = mk_eq_refl(ctx, lhs);
```

Comment:
An important reason why we want to keep the original lemma is because of `squeeze_simp`.  Replacing the proof by `eq.refl` will cause `squeeze_simp` to fail because it no longer knows that this lemma was used.

Also please please think of the case `symm && !is_iff`, I believe the current code calls `mk_eq_symm` on iff-proofs, which may or may not fail spectacularly.

Reply:
> I don't think I agree; isn't `cname` the bit that squeeze_simp looks at?

`squeeze_simp` looks at the proof term produced by simp. It doesn't know about any internal data structures.  In particular it never sees `cname`.

> I would assume that `proof` contains only an expression,

`proof` is a constant (the lemma) applied to some local constants here.  It is an expression, but it very much contains the lemma name.